Description: Provide an option to prevent synaptics from communicating the touchpad size to the xserver. 
Bug-Ubuntu: https://launchpad.net/bugs/327428

Index: xserver-xorg-input-synaptics-1.8.2/include/synaptics-properties.h
===================================================================
--- xserver-xorg-input-synaptics-1.8.2.orig/include/synaptics-properties.h
+++ xserver-xorg-input-synaptics-1.8.2/include/synaptics-properties.h
@@ -134,6 +134,9 @@
 #define SYNAPTICS_PROP_PRESSURE_MOTION_FACTOR "Synaptics Pressure Motion Factor"
 
 /* 8 bit (BOOL) */
+#define SYNAPTICS_PROP_RESOLUTION_DETECT "Synaptics Resolution Detect"
+
+/* 8 bit (BOOL) */
 #define SYNAPTICS_PROP_GRAB "Synaptics Grab Event Device"
 
 /* 8 bit (BOOL), 1 value, tap-and-drag */
Index: xserver-xorg-input-synaptics-1.8.2/man/synaptics.man
===================================================================
--- xserver-xorg-input-synaptics-1.8.2.orig/man/synaptics.man
+++ xserver-xorg-input-synaptics-1.8.2/man/synaptics.man
@@ -427,6 +427,18 @@ The gesture is enabled by default and ca
 TapAndDragGesture option to false. Property: "Synaptics Gestures"
 .
 .TP
+.BI "Option \*ResolutionDetect\*q \*q" boolean \*q
+Allow or prevent the synaptics driver from reporting the size of the
+touchpad to the X server.  The X server normally uses this information
+to scale movements so that touchpad movement corresponds visually to
+mouse cursor movements on the screen.  However, in some rare cases where
+the touchpad height/width ratio is significantly different from the
+laptop, it can cause the mouse cursor to skip pixels in the X or Y axis.
+This option allows disabling this scaling behavior, which can provide
+smoother mouse movement in such cases.
+Property: "Synaptics Resolution Detect"
+.
+.TP
 .BI "Option \*qVertResolution\*q \*q" integer \*q
 Resolution of X coordinates in units/millimeter. The value is used
 together with HorizResolution to compensate unequal vertical and
@@ -881,6 +893,10 @@ FLOAT, 2 values, speed, friction.
 FLOAT, 2 values, min, max.
 
 .TP 7
+.BI "Synaptics Resolution Detect"
+8 bit (BOOL).
+
+.TP 7
 .BI "Synaptics Grab Event Device"
 8 bit (BOOL).
 
Index: xserver-xorg-input-synaptics-1.8.2/src/properties.c
===================================================================
--- xserver-xorg-input-synaptics-1.8.2.orig/src/properties.c
+++ xserver-xorg-input-synaptics-1.8.2/src/properties.c
@@ -85,6 +85,7 @@ Atom prop_palm_dim = 0;
 Atom prop_coastspeed = 0;
 Atom prop_pressuremotion = 0;
 Atom prop_pressuremotion_factor = 0;
+Atom prop_resolution_detect = 0;
 Atom prop_grab = 0;
 Atom prop_gestures = 0;
 Atom prop_capabilities = 0;
@@ -344,6 +345,10 @@ InitDeviceProperties(InputInfoPtr pInfo)
         InitFloatAtom(pInfo->dev, SYNAPTICS_PROP_PRESSURE_MOTION_FACTOR, 2,
                       fvalues);
 
+    prop_resolution_detect =
+        InitAtom(pInfo->dev, SYNAPTICS_PROP_RESOLUTION_DETECT, 8, 1,
+                 &para->resolution_detect);
+
     prop_grab =
         InitAtom(pInfo->dev, SYNAPTICS_PROP_GRAB, 8, 1,
                  &para->grab_event_device);
@@ -746,6 +751,12 @@ SetProperty(DeviceIntPtr dev, Atom prope
         para->press_motion_min_factor = press[0];
         para->press_motion_max_factor = press[1];
     }
+    else if (property == prop_resolution_detect) {
+        if (prop->size != 1 || prop->format != 8 || prop->type != XA_INTEGER)
+            return BadMatch;
+
+        para->resolution_detect = *(BOOL*)prop->data;
+    }
     else if (property == prop_grab) {
         if (prop->size != 1 || prop->format != 8 || prop->type != XA_INTEGER)
             return BadMatch;
Index: xserver-xorg-input-synaptics-1.8.2/src/synaptics.c
===================================================================
--- xserver-xorg-input-synaptics-1.8.2.orig/src/synaptics.c
+++ xserver-xorg-input-synaptics-1.8.2/src/synaptics.c
@@ -742,6 +742,7 @@ set_default_parameters(InputInfoPtr pInf
         xf86SetIntOption(opts, "PressureMotionMinZ", pressureMotionMinZ);
     pars->press_motion_max_z =
         xf86SetIntOption(opts, "PressureMotionMaxZ", pressureMotionMaxZ);
+    pars->resolution_detect = xf86SetBoolOption(opts, "ResolutionDetect", TRUE);
 
     pars->min_speed = xf86SetRealOption(opts, "MinSpeed", 0.4);
     pars->max_speed = xf86SetRealOption(opts, "MaxSpeed", 0.7);
@@ -1303,7 +1304,7 @@ DeviceInit(DeviceIntPtr dev)
     }
 
     /* X valuator */
-    if (priv->minx < priv->maxx) {
+    if (priv->minx < priv->maxx && priv->synpara.resolution_detect) {
         min = priv->minx;
         max = priv->maxx;
     }
@@ -1318,7 +1319,7 @@ DeviceInit(DeviceIntPtr dev)
     xf86InitValuatorDefaults(dev, 0);
 
     /* Y valuator */
-    if (priv->miny < priv->maxy) {
+    if (priv->miny < priv->maxy && priv->synpara.resolution_detect) {
         min = priv->miny;
         max = priv->maxy;
     }
Index: xserver-xorg-input-synaptics-1.8.2/src/synapticsstr.h
===================================================================
--- xserver-xorg-input-synaptics-1.8.2.orig/src/synapticsstr.h
+++ xserver-xorg-input-synaptics-1.8.2/src/synapticsstr.h
@@ -218,6 +218,7 @@ typedef struct _SynapticsParameters {
     int press_motion_max_z;     /* finger pressure at which maximum pressure motion factor is applied */
     double press_motion_min_factor;     /* factor applied on speed when finger pressure is at minimum */
     double press_motion_max_factor;     /* factor applied on speed when finger pressure is at minimum */
+    Bool resolution_detect;     /* report pad size to xserver? */
     Bool grab_event_device;     /* grab event device for exclusive use? */
     Bool tap_and_drag_gesture;  /* Switches the tap-and-drag gesture on/off */
     unsigned int resolution_horiz;      /* horizontal resolution of touchpad in units/mm */
Index: xserver-xorg-input-synaptics-1.8.2/tools/synclient.c
===================================================================
--- xserver-xorg-input-synaptics-1.8.2.orig/tools/synclient.c
+++ xserver-xorg-input-synaptics-1.8.2/tools/synclient.c
@@ -131,6 +131,7 @@ static struct Parameter params[] = {
     {"PressureMotionMaxZ",    PT_INT,    1, 255,   SYNAPTICS_PROP_PRESSURE_MOTION,	32,	1},
     {"PressureMotionMinFactor", PT_DOUBLE, 0, 10.0,SYNAPTICS_PROP_PRESSURE_MOTION_FACTOR,	0 /*float*/,	0},
     {"PressureMotionMaxFactor", PT_DOUBLE, 0, 10.0,SYNAPTICS_PROP_PRESSURE_MOTION_FACTOR,	0 /*float*/,	1},
+    {"ResolutionDetect",      PT_BOOL,   0, 1,     SYNAPTICS_PROP_RESOLUTION_DETECT,	8,      0},
     {"GrabEventDevice",       PT_BOOL,   0, 1,     SYNAPTICS_PROP_GRAB,	8,	0},
     {"TapAndDragGesture",     PT_BOOL,   0, 1,     SYNAPTICS_PROP_GESTURES,	8,	0},
     {"AreaLeftEdge",          PT_INT,    0, 10000, SYNAPTICS_PROP_AREA,	32,	0},
